File: JANL/Extensions/TypeExtensions.vb
Imports System.Drawing
Imports System.Reflection
Imports System.Runtime.CompilerServices
Imports System.Threading

Namespace Extensions
	Public Module TypeExtensions
		Private ReadOnly Rnd As New Random(Today.Millisecond)

		''' <summary>
		''' Помечает все строки на удаление
		''' </summary>
		''' <param name="dt"></param>
		<Extension>
		Public Sub DeleteAll(dt As DataTable)
			Dim RemoveRowsTable = dt
			Dim i = 0
			'Remove All
			While i < RemoveRowsTable.Rows.Count
				Dim currentRow = RemoveRowsTable.Rows(i)
				If (currentRow.RowState <> DataRowState.Deleted) Then
					currentRow.Delete()
				Else
					i += 1
				End If
			End While
		End Sub

		<Extension>
		Public Function PickRandom(Of T)(L As IEnumerable(Of T)) As T
			If L.Count = 0 Then Return Nothing
			Return L(Rnd.Next(L.Count))
		End Function

		<Extension>
		Public Function PostDEC(ByRef x As Integer) As Integer
			Dim tmp = x
			Interlocked.Decrement(x)
			Return tmp
		End Function

		<Extension>
		Public Function PostINC(ByRef x As Integer) As Integer
			Dim tmp = x
			Interlocked.Increment(x)
			Return tmp
		End Function

		<Extension>
		Public Function PreDEC(ByRef x As Integer) As Integer
			Return Interlocked.Decrement(x)
		End Function

		<Extension>
		Public Function PreINC(ByRef x As Integer) As Integer
			Return Interlocked.Increment(x)
		End Function

		<Extension>
		Public Function ToIcon(Bitmap As Bitmap) As Icon
			Using I = Icon.FromHandle(Bitmap.GetHicon)
				'Copy icon to change owner
				Dim Icon = DirectCast(I.Clone(), Icon)
				WinAPI.DestroyIcon(I.Handle)
				Return Icon
			End Using
		End Function

		<Extension>
		Public Function Truncate(value As String, maxLength As Integer) As String
			If (String.IsNullOrEmpty(value)) Then Return value
			Return If(value.Length <= maxLength, value, value.Substring(0, maxLength))
		End Function

		<Extension>
		Public Function IsOverride(M As MethodInfo) As Boolean
			Return M.GetBaseDefinition().DeclaringType <> M.DeclaringType
		End Function

		<Extension>
		Public Function IsOverride(T As Type, Name As String) As Boolean
			Return T.GetMethod(Name).IsOverride()
		End Function

#Region "Conversion"

		''' <summary>
		''' Преобразует DataTable в список объектов типа Т
		''' </summary>
		''' <typeparam name="T"></typeparam>
		''' <param name="DT"></param>
		Public Function ToObject(Of T As {Class, New})(DT As DataTable) As List(Of T)
			Dim D As Dictionary(Of String, String) = DT.Columns.Cast(Of DataColumn)().ToDictionary(Function(k) k.ColumnName.ToLower, Function(v) v.ColumnName)
			Dim Props = GetType(T).GetProperties(BindingFlags.Instance Or BindingFlags.Public).ToList()

			Dim Items = New List(Of T)()
			For Each R As DataRow In DT.Rows
				Dim item = New T()
				For Each prop In Props
					If Not D.ContainsKey(prop.Name.ToLower()) Then Continue For
					Dim Collumn = D(prop.Name.ToLower())
					If Not R.IsNull(Collumn) Then prop.SetValue(item, R(Collumn), Nothing)
				Next
				Items.Add(item)
			Next

			Return Items
		End Function

		''' <summary>
		''' Преобразует DataRow в объект типа <typeparamref name="T"/>
		''' </summary>
		''' <typeparam name="T"></typeparam>
		''' <param name="R"></param>
		Public Function ToObject(Of T As {Class, New})(R As DataRow) As T
			Dim D As Dictionary(Of String, String) = R.Table.Columns.Cast(Of DataColumn)().ToDictionary(Function(k) k.ColumnName.ToLower, Function(v) v.ColumnName)
			Dim Props = GetType(T).GetProperties(BindingFlags.Instance Or BindingFlags.Public).ToList()
			Dim Item = New T()
			For Each prop In Props
				If Not D.ContainsKey(prop.Name.ToLower()) Then Continue For
				Dim Collumn = D(prop.Name.ToLower())
				If Not R.IsNull(Collumn) Then prop.SetValue(Item, R(Collumn), Nothing)
			Next
			Return Item
		End Function

#End Region

	End Module
End Namespace

File: JANL/Extensions/TypeExtensions.cs
using JANL.SQL;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Reflection;

namespace JANL.Extensions
{
    /// <summary>
    /// Расширения для базовых типов
    /// </summary>
    public static class TypeExtensions
    {
        private static readonly Random rnd = new Random();

        /// <summary>
        /// Ограничивает значение диапазоном
        /// </summary>
        /// <param name="value"></param>
        /// <param name="min">Минимум</param>
        /// <param name="max">Максимум</param>
        public static T Clamp<T>(this T value, T min, T max) where T : struct, IComparable<T>
        {
            if (value.CompareTo(min) < 0) { return min; }
            if (value.CompareTo(max) > 0) { return max; }
            return value;
        }

        /// <summary>
        /// Случайный элемент из перечисления
        /// </summary>
        public static T PickRandom<T>(this IEnumerable<T> list) => list.Count() == 0 ? default : list.ElementAt(rnd.Next(list.Count()));

        /// <summary>
        /// Случайный элемент из коллекции
        /// </summary>
        public static T PickRandom<T>(this ICollection<T> collection) => PickRandom<T>(collection, false);

        /// <summary>
        /// Случайный элемент из коллекции
        /// </summary>
        /// <param name="collection"></param>
        /// <param name="remove">Удалить элемент из коллекции</param>
        public static T PickRandom<T>(this ICollection<T> collection, bool remove)
        {
            if (collection.Count == 0) { return default; }
            var element = collection.ElementAt(rnd.Next(collection.Count));
            if (remove) { collection.Remove(element); }
            return element;
        }

        #region SQL

        /// <summary>
        /// Получить тип CRL
        /// </summary>
        /// <param name="type"></param>
        /// <returns></returns>
        public static Type GetCRLType(this SqlDbType type) => SQLHelper.GetCRLType(type);

        /// <summary>
        /// Получить тип SQL
        /// </summary>
        /// <param name="type"></param>
        /// <returns></returns>
        public static SqlDbType GetSQLType(this Type type) => SQLHelper.GetDBType(type);

        #endregion SQL

        #region IsOverride

        /// <summary>
        /// Переопределён ли метод
        /// </summary>
        /// <param name="MI"></param>
        /// <returns></returns>
        public static bool IsOverride(this MethodInfo MI) => MI.GetBaseDefinition().DeclaringType != MI.DeclaringType;

        /// <summary>
        /// Переопределёно ли свойство
        /// </summary>
        /// <param name="PI"></param>
        /// <returns></returns>
        public static bool IsOverride(this PropertyInfo PI) => PI.GetMethod.GetBaseDefinition().DeclaringType != PI.DeclaringType;

        /// <summary>
        /// Переопределён ли метод или свойство
        /// </summary>
        /// <param name="T">Тип</param>
        /// <param name="Name">Имя метода или свойства</param>
        /// <returns></returns>
        public static bool IsOverride(this Type T, string Name)
        {
            var Members = T.GetMember(Name);
            if (Members.Count() == 0) { return false; }
            var M = Members.First();
            switch (M.MemberType)
            {
                case MemberTypes.Method:
                    return IsOverride((MethodInfo)M);

                case MemberTypes.Property:
                    return IsOverride((PropertyInfo)M);

                default:
                    return false;
            }
        }

        #endregion IsOverride
    }
}